### DISCUSSION TITLE
[RFC] 提供 Local History 支持

### DISCUSSION THREAD

**Discussion by Ricbet** [opensumi/core#2478] [RFC]
### 背景
在 IDE 里进行编码可能会发生因为误操作等原因导致代码丢失的情况，比如误操作 git reset、不在暂存区的文件撤销/回退栈丢失的问题等。<br />为了解决这个问题，我们打算在 OpenSumi 里内置实现一个 Local History 功能，它可以保存代码的历史操作，并能够方便地进行版本比较和恢复。

与 Git 版本控制系统不同，Local History 可以记录项目的所有变更操作，包括代码编辑、保存、重命名、撤销/回退、来自 workspace edit 的操作等，并允许你回滚到文件的任何状态。即使你的项目还没启用 Git 版本控制，他也可以成为你的个人版本控制系统。

但与 Git 不同的是，Local History 需要设定存储的大小和时间，所以他并不是 Git 的替代品，而是作为一种兜底的方案来帮助你找回曾经丢失的代码

#### WebStorm
https://www.jetbrains.com/help/webstorm/local-history.html#restore-changes-from-local-history

#### VS Code
https://code.visualstudio.com/updates/v1_66#_local-history

### 功能清单

- 单文件级别的代码操作历史记录，记录以下操作
   - 文件创建
   - 文件替换
   - 代码保存
   - 文件重命名
   - 文件移动
   - Undo / Redo
   - 来自 Workspace Edit 的操作（例如重构）
   - （欢迎更多补充）
- 查看本地历史记录
   - 有两种查看方式
      - 当打开某个文件时，左侧面板提供该文件的所有本地历史记录
      - 左侧面板默认展示当前工作区的所有本地历史文件列表（点击后如何查看历史的交互还没想好，前期先直接调用 command 命令打开吧）
   - 点击可查看对比 diff
- 恢复/删除操作
   - 支持恢复某个操作
   - 支持删除某个操作
- 配置相关
   - 支持配置启动开关
   - 支持配置 exclude
   - 支持配置限制每个文件存储的条目个数
   - 支持配置限制最大文件大小
   - 支持配置多少秒对同一种操作做本地存储，例 10 秒内重复执行保存操作则只记录一次（默认 10 秒）
   - 上游集成的配置项
      - 支持配置存储路径
         - windows: 待补充
         - linux：待补充
         - mac：待补充
      - 支持配置限制存储的总大小
      - (欢迎更多补充)
- 插件 API
   - registerTimelineProvider 实现

#### 关于未保存文件的历史记录处理
对于未保存的文件需要单独做 backup 处理，既未保存时会在某个缓存路径创建临时文件并存储该文档的内容，保存后再将这个文件删除<br />所以对于未保存的文件不记录在 Local History 的范围内，这块需要另开 issue 单独实现

### Actions
（待补充）

### 如何测试？
（待补充）

**Comment by erha19** [comment]
这里就涉及到了 TimeLine 面板的实现，相关 Issue https://github.com/opensumi/core/issues/2424.

**Reply by Ricbet** [reply]
对，但 timeline 面板和 local history 功能不是耦合的，可以分别实现

**Comment by Ricbet** [comment]
结合 https://github.com/opensumi/core/issues/2480 来给未保存的文件提供备份存储，保存后则进入到 Local History 的操作记录中

**Reply by Ricbet** [reply]
内部已有 [`IDocPersistentCacheProvider`](https://github.com/opensumi/core/blob/58b998d9e1/packages/editor/src/common/doc-cache.ts#L56) 提供文档缓存支持的接口

web 端的默认实现是 https://github.com/opensumi/core/blob/58b998d9e1f721928f576579f16ded46b7505e84/packages/editor/src/browser/doc-cache/local-storage-cache.ts#L11

electron 本地端还未默认实现
